sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala

```diff
@@ -404,6 +407,19 @@ class JacksonParser(
       }
     }
 
+    // When the input schema is setting to `nullable = false`, make sure the field is not null.
+    var index = 0
+    while (badRecordException.isEmpty && !skipRow && index < schema.length) {
```

Comment:
Can we fix `parseJsonToken` instead by passing `nullable` bool flag?

Reply:
Let me try.

Reply:
As the missing field won't call `fieldConverter`, it seems it's hard to do this? https://github.com/apache/spark/blob/2537fe8cbaf49070137d4b5bc39af078b306c4c8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala#L404-L419

Reply:
So there are 2 cases that can lead to null result:
1. the JSON value is null
2. the JSON field is missing

Reply:
and for malformed fields, we also set the value to null and put the JSON string to a special column.

Reply:
We may not return the JSON string, the reason show in [test](https://github.com/apache/spark/pull/33212/files#r666705223)